File: c++/validAnagram#242/try1.cpp
// This is technically the generic optimal solution to the problem.
// There is another more optimal solution for this specific problem.

class Solution {
public:
    bool isAnagram(string s, string t) 
    {
        if(s.size() != t.size())
        {
            return false; 
        }    
        unordered_map<char, int> sMap;
        unordered_map<char, int> tMap;
        for(int i = 0; i < s.size(); i++)
        {
            sMap[s[i]]++;
            tMap[t[i]]++;
        }
        if(sMap.size() != tMap.size())
        {
            return false;
        }

        for(int i = 0; i < tMap.size(); i++)
        {
            if(sMap[i] != tMap[i])
            {
                return false;
            }
        }
        return true;
    }
};

//time complexity O(N)
//space complexity O(N)

File: python/subsets#78/try1.py
# This solution is not great but technically has optimal complexity for time
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        s = set()

        def dfs(i, curSet):
            s.add(tuple(curSet))

            while(i < len(nums)):

                dfs(i + 1, curSet[:])
                curSet.append(nums[i])
                dfs(i + 1, curSet[:])
                i += 1

        dfs(0, [])

        return list(s)

# Time: O(N*2^N)
# Space: O(N*2^N)
